### ISSUE TITLE
web: VoiceOver does not announce AlertDialog when shown

### ISSUE THREAD

**Issue by vsomayaji** [flutter/flutter#97750]
## Steps to Reproduce

1. Execute the code sample on web. 
2. Turn on VoiceOver.
3. Click "Show Dialog".

**Expected results:**

Focus changes to the shown `AlertDialog` and VoiceOver reads either the title or the button.

There is an example of a dialog that is read correctly by VoiceOver at https://adrianroselli.com/2020/10/dialog-focus-in-screen-readers.html.

**Actual results:**

VoiceOver does not read anything. Clicking **Tab** changes focus to the button and then VoiceOver reads it.

<details>
<summary>Code sample</summary>

https://dartpad.dev/?id=4400d4a07c7d0e782a629ed952af7b07
</details>

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, 2.10.0, on macOS 11.6.2 20G314 darwin-x64, locale en-US)
    • Flutter version 2.10.0 at /Users/vsomayaji/Dev/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5f105a6ca7 (2 days ago), 2022-02-01 14:15:42 -0800
    • Engine revision 776efd2034
    • Dart version 2.16.0
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/vsomayaji/Library/Android/sdk
    • Platform android-31, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by darshankawar** [comment]
@vsomayaji 
As I understand this behavior, focusing directly on the `AlertDialog` reads it as `group` and then focussing on dialog title and button title reads it properly as below:

<img width="982" alt="Screenshot 2022-02-04 at 1 44 36 PM" src="https://user-images.githubusercontent.com/67046386/152495360-1e115a41-916e-4a40-b0de-c865871699b9.png">
<img width="985" alt="Screenshot 2022-02-04 at 1 44 55 PM" src="https://user-images.githubusercontent.com/67046386/152495367-a2f0d7c4-b665-4f97-a593-066b85e4fe7f.png">

Per expected result, focussing on `AlertDialog` should read it the title and button instead of `group`. Is this a correct interpretation ?

**Comment by vsomayaji** [comment]
@darshankawar thanks for the quick response! That's correct. Our expectation was that by showing the `AlertDialog`, at a minimum, the button would be focused and read. But, in testing on VoiceOver, JAWS and NVDA, there is no indication that a dialog has been shown.

I updated the code sample with a workaround that appears to do the trick. If we add `autofocus: true` to the button, exclude the title from the semantics tree but re-add it with the button, we seem to get the behavior we want.

```dart
          builder: (BuildContext context) => AlertDialog(
            // Workaround: Exclude the title from the semantics tree.
            title:
                const ExcludeSemantics(child: Text('This title is announced')),
            actions: [
              TextButton(
                autofocus: true, // Workaround: Autofocus the node.
                onPressed: () => Navigator.pop(context),
                // Workaround: Add back the title to the semantics tree.
                child: Semantics(
                  label: 'This title is announced',
                  child: const Text('So is this button'),
                ),
              ),
            ],
          ),
        ),
```

**Comment by darshankawar** [comment]
@vsomayaji 
Can you check and try solution mentioned in [this](https://github.com/flutter/flutter/issues/76108#issuecomment-780723834) issue comment and see if it works for your case ?

**Comment by vsomayaji** [comment]
@darshankawar thanks but no luck. 😦 Neither [`SemanticsProperties.sortKey`](https://api.flutter.dev/flutter/semantics/SemanticsProperties/sortKey.html) nor [`SemanticsProperties.focused`](https://api.flutter.dev/flutter/semantics/SemanticsProperties/focused.html) resulted in the [`AlertDialog`](https://api.flutter.dev/flutter/material/AlertDialog-class.html) getting the focus.

The workaround I posted (using [`autofocus`](https://api.flutter.dev/flutter/material/ButtonStyleButton/autofocus.html) on the button and combining the semantics of the title and button is enough to unblock us, so this can be closed.

From a screen reader perspective, it would be ideal if the `AlertDialog` on web had the ARIA `alertdialog` semantics as described at https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/alertdialog_role. Should I submit a feature request for that?

**Comment by darshankawar** [comment]
> Should I submit a feature request for that?

Yes. 
Will close this as you have already provided a solution.
